### PR TITLE
Store per-device playback preferences in localStorage

### PR DESCRIPTION
**Pull request by jibanez74**
## Why

`users` carried four playback preference columns — `preferred_hls_profile`, `download_mbps`, `preferred_audio_language`, `preferred_subtitle_language` — that **no server code ever read**. They were written by the playback settings endpoint and handed straight back to the browser, using the database as a slow, shared substitute for local storage.

They are also keyed by the wrong thing. One account streams from a phone on LTE, a laptop, and a TV on ethernet; a preferred profile or download speed that is right on one is wrong on the next.

## What changed

**Preferences move to localStorage**, scoped per user *and* per device (`igloo-playback-prefs:<userId>`) — a browser profile can be shared by several people, and language preferences are personal. Pre-production, so nothing is migrated: the columns and their two sqlc queries are deleted outright.

**The endpoint keeps its server-owned half** — the transcode profile catalog, the server upload cap, and `hardware_acceleration_device`, which `hls_session.go:785` genuinely consumes. The `PUT` now carries only the two admin fields, is gated by `RequireAdmin` instead of an in-handler field check, needs no transaction (one table is written now), and echoes the same envelope the `GET` returns. The handler drops 394 → 199 lines and three OpenAPI schemas disappear.

**Settings → Playback splits by ownership**: two "this device" cards that apply instantly, plus an admin-only "Server" card carrying the only save bar. Non-admins make zero API writes from this tab.

**Sanitization is the new boundary.** localStorage is user-writable and the server no longer validates these values, so the reader coerces every field (unknown profile ids, out-of-range speeds, malformed language codes). `getDefaultPlaybackSettings` still re-checks the mode against `availableModes`, and the HLS endpoints still validate the profile server-side.

## Behaviour win

Only the download-speed → recommended-profile path still waits on the network. A stored profile resolves playback with no request at all, so there is no longer a flash of the wrong default audio/subtitle track while preferences load.

## Notes for review

- `RETURNING *` on `users` is **not** a problem for existing databases: sqlc expands it to an explicit column list at generation time. Verified against the real dev DB — the four leftover columns are inert, no `ALTER` needed.
- The recommendation now reflects the *saved* server cap rather than an admin's unsaved draft. Arguably more honest; called out because the e2e assertion moved.
- `docs/design-system.md` §3.7 documents the new instant-apply pattern (scope in the card description + `LiveAnnouncer` confirmation, never mixed with a save bar in one card).

## Verification

- `make check` (test-openapi, check-openapi, test-server, lint-web, test-web, build-web) — green
- 699 web unit tests, incl. a new 29-case suite for the storage module (sanitization, per-user isolation, snapshot identity for `useSyncExternalStore`)
- Playwright: `playback-settings`, `movie-details`, `movie-extra-videos`, `account/libraries/user/general-settings` — all pass
- Manual smoke test against a copy of the real dev library: boots clean, `GET` returns the trimmed payload, partial `PUT` preserves the omitted field, non-admin gets 200 on `GET` / 403 on `PUT`, and user writes still work against the stale schema

Net **−589 lines**.

---

# Follow-up: review fixes (`4991c0b0`)

A review of the migration above raised three findings. Two were valid; the third was real but aimed slightly off target.

## Cross-tab clobbering (fixed)

Each tab cached its own snapshot and only invalidated it on its own writes, so a tab merging a patch against a stale snapshot wrote the whole object back and silently cleared what another tab had just set. The reader now refreshes on the `storage` event — and because that event is delivered *asynchronously*, the writer bases its merge on storage rather than on the cached snapshot. **That second part is what actually fixes the data loss**; the listener only keeps the UI live.

Reproduced before fixing: the new two-tab Playwright test fails on the old code with the audio language clobbered to `null`.

## Storage failures reported as durable saves (fixed)

localStorage can refuse a write (private browsing, quota) and the card announced "Saved on this device" anyway. That matters more here than in the theme module: the Save bar is gone, so the announcement is the *only* confirmation. The writer now returns whether it persisted, the announcement says the value applies for this session only, and both device cards carry a standing notice for as long as the condition lasts — a live announcement is one-shot, the condition is not. `docs/design-system.md` §3.7 updated to make this part of the contract.

## Concurrent partial `PUT`s (fixed, but re-scoped)

Reported as a playback bug. It isn't: the same read-modify-write runs in the general and library settings writers, it predates this branch, and the shipped client always sends both fields. The review also missed the sharper defect on the same line — **`app.Settings` was an unsynchronized pointer write**, read concurrently by the static and avatar handlers, both scanners, and every HLS session. A genuine data race the detector flags.

So it is fixed at the right scope: the field is unexported behind a `RWMutex` with `CurrentSettings`/`SetSettings`, and all three writers run their read-modify-write inside `withSettingsWrite`, which holds the lock across load → update → cache refresh. Filesystem validation stays outside the lock. Serializing rare admin writes costs nothing, and this closes the reported finding as a side effect rather than papering over it in one handler.

## Dead code removed

Settings are never nil once the application exists — `InitSettings` runs in `NewApplication` and fails the boot otherwise — so the nil branches claiming otherwise are gone, along with the database fallback they guarded. `mapPlaybackSettingsResponse` and `hardwareAccelerationDeviceOrDefault` now take the row **by value**, so nil is not representable rather than merely unchecked.

Only the test helpers ever produced that state. They now boot settings the way production does, which is a deliberate change: tests were pinning behaviour in a state the running server cannot reach. Fallout — one subtest existed solely to drive `settings = nil` (deleted; the blank-column case still covers the real fallback), three `InitSettings` tests needed an explicit empty table, and `setupTestApp` needed an absolute `StaticDir` so avatar writes stop landing in the package directory.

## Refactors

The route component is lazily code-split and cannot render outside a router, which left the device cards untestable. They move to `components/settings/` alongside `DevicesCard`, taking with them the select markup that was repeated three times (now one `PreferenceSelect`) and `parseMbpsInput`, which was duplicated (now in `lib/playback.ts`).

## Verification

- `make check` exit 0 — **710** web unit tests, all Go packages, build clean
- `make test-openapi` ok — no API contract change; the "absent fields keep their value" contract is now actually honoured under concurrency rather than reworded
- `go vet` clean; **`go test -race ./cmd/api` clean** — this is the real gate on the mutex work
- Playwright `playback-settings`: 4/4, including the new cross-tab spec
- `bun run doctor`: no findings on any touched file

Each of the three regression tests was confirmed to **fail against the unfixed code** — the Go concurrency test (2 of 30 runs without the lock), the vitest merge test, and the Playwright cross-tab test.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Playback preferences now save immediately per browser and device, without a Save bar.
  * Device preferences include profile, download speed, audio language, and subtitle language.
  * Changes are announced for improved accessibility.
  * Server playback settings remain available, with updates restricted to administrators.

* **Bug Fixes**
  * Improved validation and handling of invalid or unavailable preferences.
  * Playback recommendations now combine device preferences with server settings reliably.
  * Added clear warnings when preferences can only be saved for the current session.

* **Documentation**
  * Clarified the distinction between device-scoped preferences and server-owned settings.

### CODE REVIEW

File: docs/openapi.json [hunk 1]
```diff
@@ -3166,15 +3166,15 @@
...
+        "summary": "Update server playback settings",
+        "description": "Admin-only. Replaces the server-wide upload bandwidth cap and hardware acceleration device. Per-device playback preferences are not stored on the server; they live in the client's local storage.",
         "operationId": "updatePlaybackSettings",
         "requestBody": {
           "$ref": "#/components/requestBodies/UpdatePlaybackSettingsRequest"
         },
         "responses": {
           "200": {
-            "$ref": "#/components/responses/UpdatePlaybackSettingsResponse"
+            "$ref": "#/components/responses/PlaybackSettingsResponse"
```
[the remark this answers is not in the document]
Comment (jibanez74) [reply docs/openapi.json:3177]: Fixed in 82784d06. Description now reads "Updates ... ; fields absent from the body keep their current value", and `web/src/types/openapi.gen.ts` was regenerated with `bun run generate:openapi`.

File: server/cmd/api/playback_settings_handler_test.go [hunk 2]
```diff
@@ -561,6 +327,9 @@ func TestUpdatePlaybackSettings_RegularUserCannotUpdateServerUploadCap(t *testin
 	if !gotSettings.ServerUploadMbps.Valid || gotSettings.ServerUploadMbps.Float64 != 22 {
 		t.Fatalf("expected server_upload_mbps unchanged at 22, got valid=%v %v", gotSettings.ServerUploadMbps.Valid, gotSettings.ServerUploadMbps.Float64)
 	}
+	if gotSettings.HardwareAccelerationDevice.String == helpers.HARDWARE_ACCELERATION_DEVICE_NVIDIA {
+		t.Fatal("expected hardware device unchanged for regular user")
+	}
```
[the remark this answers is not in the document]
Comment (jibanez74) [reply server/cmd/api/playback_settings_handler_test.go]: Fixed in 82784d06. The assertion now requires `.Valid` and an exact match on `HARDWARE_ACCELERATION_DEVICE_CPU`, the value `seedServerPlaybackSettings` stores, so an empty string or any other device fails it.

File: web/e2e/mock-api-server.ts [hunk 3]
```diff
@@ -1327,54 +1272,41 @@ async function handleSettingsRoutes(
...
-      settings: {
-        preferred_profile: settings.preferred_profile,
-        download_mbps: settings.download_mbps,
-        preferred_audio_language: settings.preferred_audio_language,
-        preferred_subtitle_language: settings.preferred_subtitle_language,
-      },
-    });
+    sendSuccess(response, { settings: playbackSettingsResponse() });
     return true;
   }
```
[the remark this answers is not in the document]
Comment (jibanez74) [reply web/e2e/mock-api-server.ts:1348]: Fixed in 82784d06, with one correction to the framing: no spec passes falsely today, because `routes/_auth/settings/playback.tsx` blocks an out-of-range bandwidth before it submits and the device value comes from a fixed `Select`. This is a latent trap for a future spec rather than a live bug, so Minor rather than Major.

The mock now mirrors both Go checks, including that the real handler rejects a JSON `null` device with 400. Implemented with an `isHardwareAccelerationDevice` type guard rather than the proposed `includes` + cast, which lets the union be enforced by a check instead of assumed by an assertion.

File: web/src/hooks/useMoviePlaybackData.ts [hunk 4]
```diff
@@ -79,16 +81,24 @@ export function useMoviePlaybackData({
...
+  );
+  // Device preferences are synchronous, so the only thing still worth waiting
+  // for is the server catalog -- and getDefaultPlaybackSettings consults it on
+  // exactly one path: no preferred profile, but a download speed to size one
+  // against. Everything else (audio/subtitle language) resolves immediately.
+  const needsServerCatalog =
+    devicePrefs.preferredProfile === null && devicePrefs.downloadMbps !== null;
   const playbackPreferencesReady =
-    !authUserPending && (user === null || !playbackSettingsPending);
+    !authUserPending && (!needsServerCatalog || !playbackSettingsPending);
```
[the remark this answers is not in the document]
Comment (jibanez74) [reply web/src/hooks/useMoviePlaybackData.ts]: Confirmed and fixed in 82784d06 -- `getDefaultPlaybackSettings` does fall through to the download-speed branch when the stored profile is not in `availableModes`, and readiness reported true for it.

Applied a narrower gate than proposed. `needsServerCatalog = downloadMbps !== null` also waits when the stored profile *does* apply, which gives up the "a stored profile needs nothing from the network" property this PR was built around (pinned by "resolves immediately from a stored profile while the catalog is still loading"). Instead the readiness computation moved below `availableModes` and gates on whether the stored profile actually applies:

```ts
const storedProfileApplies =
  devicePrefs.preferredProfile !== null &&
  (availableModes?.some((m) => m.id === devicePrefs.preferredProfile) ?? false);
const needsServerCatalog =
  !storedProfileApplies && devicePrefs.downloadMbps !== null;
```

Covered by a new test, "waits for the catalog when the stored profile is unavailable for the file", verified to fail against the old condition.

File: server/cmd/api/music_scanner_test.go [hunk 5]
```diff
@@ -930,7 +930,7 @@ func TestRunMusicScanDoesNotClearSpotifyRuntimeCache(t *testing.T) {
 	spotifyStub := &musicScannerSpotifyStub{}
 	app.Ffprobe = &countingMusicScannerFfprobe{result: testMusicMetadata()}
 	app.Spotify = spotifyStub
-	app.Settings = &database.Setting{
+	app.settings = &database.Setting{
```
[the remark this answers is not in the document]
Comment (jibanez74) [reply server/cmd/api/music_scanner_test.go]: Fixed in 82784d06, at wider scope than reported. The field is written directly at 13 sites across 6 files, not 2 -- also `hls_session_test.go:819,828`, `main_test.go:1667`, `movies_scanner_test.go:194,225,1289` and `playback_settings_handler_test.go:65`. All now go through `SetSettings`.

Three further sites mutated the row *in place* (`app.settings.ServerUploadMbps = ...` and two in `settings_handler_test.go`), which the field comment forbids more specifically than a replacement does -- `CurrentSettings` hands out a pointer callers read without the lock. Those now copy, mutate the copy, and `SetSettings` it.

The ~85 plain reads are left as they are: they are synchronous assertions after the handler returned, so there is nothing to serialize, and converting them would be churn.

File: web/src/components/settings/DevicePlaybackCards.tsx [hunk 6]
```diff
@@ -0,0 +1,398 @@
...
+  const downloadInvalid = isDownloadSpeedOutOfRange(pendingDownload);
+
+  const handleDownloadMbpsChange = (value: string) => {
+    setDownloadText(value);
+    const parsed = parseMbpsInput(value);
+    if (isDownloadSpeedOutOfRange(parsed)) return;
+    if (parsed === prefs.downloadMbps) return;
+
+    write({ downloadMbps: parsed });
+  };
```
[the remark this answers is not in the document]
Comment (jibanez74) [reply web/src/components/settings/DevicePlaybackCards.tsx:202]: Fixed in 82784d06. `handleDownloadMbpsCommit` clears `downloadText` after a valid commit, so the field renders from `prefs`.

One note on the two effects you list: the leading-zero case is real in a browser but not reproducible in jsdom -- `user.type(input, "0025")` on `type="number"` lands as `"25"`, and `toHaveValue` would coerce it anyway. The cross-tab half is testable and is what the new test "follows another tab's value after committing its own" pins, via a dispatched `StorageEvent`.

File: web/src/components/settings/DevicePlaybackCards.tsx [hunk 7]
```diff
@@ -0,0 +1,398 @@
...
+  // from the standing flag rather than a fresh return value.
+  const handleDownloadMbpsCommit = () => {
+    if (downloadInvalid) return;
+    announce(
+      pendingDownload === null
+        ? "Download speed cleared."
+        : `Download speed set to ${pendingDownload} Mbps.`,
+      !persistenceFailed,
+    );
+  };
```
[the remark this answers is not in the document]
Comment (jibanez74) [reply web/src/components/settings/DevicePlaybackCards.tsx:221]: Confirmed and fixed in 82784d06 -- an early `if (downloadText === null) return;` in `handleDownloadMbpsCommit`. New test: "confirms nothing when the field is blurred without an edit". It has to wait out `LiveAnnouncer`'s 100ms delay before asserting absence, or it passes either way.

File: web/src/lib/playback-preferences.ts [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+  }
+
+  listeners.add(listener);
+  return () => {
+    listeners.delete(listener);
+    if (listeners.size === 0 && typeof window !== "undefined") {
+      window.removeEventListener("storage", handleStorageEvent);
+    }
+  };
+}
```
[the remark this answers is not in the document]
Comment (jibanez74) [reply web/src/lib/playback-preferences.ts:245]: Fixed in 82784d06, but not with the proposed patch -- a bare `snapshots.clear()` introduces a different data-loss bug. A user in `unpersisted` has values that exist *only* in `snapshots`, because `setItem` threw; clearing the entry makes the next `setDevicePlaybackPreferences` merge against `read()` (line 187) and silently drop them. The unsubscribe now skips those users:

```ts
for (const cachedUserId of snapshots.keys()) {
  if (!unpersisted.has(cachedUserId)) {
    snapshots.delete(cachedUserId);
  }
}
```

Two tests: "re-parses storage for a read after the last subscriber leaves" (fails without the invalidation) and "keeps a session-only value across the last unsubscribe" (fails against the verbatim suggestion).
